### PR TITLE
Add type hint for drive control module

### PR DESCRIPTION
**Pull request by tachikoma-li**
Add type hint

For passing `mypy` test, some refactoring is done for checking the inputs. The main purposes of these checking are:
1. set proper default values when inputs are all ``None``
2. make sure all non ``None`` inputs have the same shape
3. set the proper values for ``None`` input

Note that type hint is not added for those properties as I am planning to rearrange those checkings first. Will do in another PR.

### CODE REVIEW

File: qctrlopencontrols/driven_controls/driven_control.py [hunk 1]
```diff
@@ -65,80 +67,59 @@ class DrivenControl:
...
-                durations = np.ones((valid_input_length,))
-
-        self.rabi_rates = rabi_rates
-        self.azimuthal_angles = azimuthal_angles
-        self.detunings = detunings
-        self.durations = durations
+        # check if all non-None inputs have the same length
+        input_lengths = {
+            len(v)
+            for v in [rabi_rates, azimuthal_angles, detunings, durations]
```
Comment (charmasaur) [inline qctrlopencontrols/driven_controls/driven_control.py:89]: We also used to handle the case where these were multi-dimensional (we'd check the lengths of the flattened inputs). I guess that could be relevant if the user passes `[[1,2,3]]` or something... It's a little inefficient, but the simplest way I can think of to handle that case is:
 - change the `len(v)` to `len(np.array(v).size)`
 - call `.flatten()` when creating the arrays on 116--119

Comment (tachikoma-li) [reply qctrlopencontrols/driven_controls/driven_control.py:89]: SG.

> We also used to handle the case where these were multi-dimensional (we'd check the lengths of the flattened inputs). I guess that could be relevant if the user passes `[[1,2,3]]` or something... It's a little inefficient, but the simplest way I can think of to handle that case is:
> 
>     * change the `len(v)` to `len(np.array(v).size)`
IIUC, `np.array(v).size` would be fine for handling size?
>     * call `.flatten()` when creating the arrays on 116--119
SG

Comment (charmasaur) [reply qctrlopencontrols/driven_controls/driven_control.py:89]: > SG.
> 
> > We also used to handle the case where these were multi-dimensional (we'd check the lengths of the flattened inputs). I guess that could be relevant if the user passes `[[1,2,3]]` or something... It's a little inefficient, but the simplest way I can think of to handle that case is:
> > ```
> > * change the `len(v)` to `len(np.array(v).size)`
> > ```
> 
> IIUC, `np.array(v).size` would be fine for handling size?

D'oh.

Comment (tachikoma-li) [reply qctrlopencontrols/driven_controls/driven_control.py:89]: Done. The only issue I have now is that the docstring says all inputs are 1-D. Any reason why we want to support the multi-dimensional case here?

Comment (charmasaur) [reply qctrlopencontrols/driven_controls/driven_control.py:89]: It's probably not really necessary, I think it's generally just good if wherever possible we can avoid changing functionality (in a backwards incompatible way). In this case we can manage that without too much trouble, so I think it's worth it.

**Pull request merged**